### DISCUSSION TITLE
Creating Custom  Datasets

### DISCUSSION THREAD

**Discussion by mprithiv** [pyg-team/pytorch_geometric#3872] [Q&A]
Hello,

I am working on creating a graph embedding using a datasets as buyers.csv, sellers.csv and bought.csv files. The buyers will have a uniqueBuyerID and 4 other features. The sellers data will have sellerUniqueID and 3 other features. The bought.csv will have uniqueBuyerID, uniqueSellerID, amount, date. I am trying to create a heterogeneous graph and train an unsupervised embedding with RGCN. I am using one of the columns in buyer.csv as a label. I tried using the movie rating example and create a graph. What would be the numpy array for both the nodes would look like? How would I include the node and edge features?

Nodes: buyers, sellers
Edges: amount, date

While performing RandomNodeSplit and RandomLinkSplit, how to make sure the split contains information across all the nodes and edges? What are the results of applying a simple GCN over this graph rather than using RGCN (do not see any implementation of that in PyG) provided I have a heterogeneous bipartite graph.

**Comment by rusty1s** [comment]
That looks correct to me. Buyers and sellers would be a node type, in which each node/row has 3 or 4 features, i.e., `buyer.x` has shape `[num_buyers, 4]` and `seller.x` has shape `[num_sellers, 3]`. Since you want to predict a label of a buyer, you can use `RandomNodeSplit` to further split the data.

Since the feature dimensionality of buyers and sellers is different, you will have to use some heterogeneous GNN to process the data. As an alternative, you can try to map the features of buyers and sellers first into a unified embedding space before using a homogeneous GNN for further processing. It's hard to tell which approach works better in practice in your case.

**Reply by mprithiv** [reply]
Interesting. Any approaches within PyG are suggested to convert heterogeneous to homogenous? Only approach I can think of is manual steps of 
1. creating buyer graph with sellerID as edge, amount and date as edge attributes
2. Creating seller graph with buyerId as edge, amount and date as edge attributes

Not sure if this is a correct representation and if encodes all the relationships between the nodes. Also, in case of multiple homogeneous graphs and embedding, when making a prediction, I assume, for buyer node, I need to use buyer embedding sand for seller, seller embedding?

**Reply by rusty1s** [reply]
There exists a [`HeteroData.to_homgeneous()`](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.HeteroData.to_homogeneous) function which might be of interest to you.

**Reply by mprithiv** [reply]
Hi 

thanks for your response. I have had a look at it before. Does this work for multiple edge types? But I wasn’t able to explicitly convert (I might be missing something here)

data.to_homogeneous()  throws error Edge storage object has not attribute edge_index
(:

**Reply by rusty1s** [reply]
Can you show me a small example to reproduce? I think currently, each edge type expects to have an `edge_index` attribute. If you do not have any edge indices for certain edge types, it's probably good to either delete the edge type or set `edge_index` to `torch.empty((2, 0), dtype=torch.long)`.

**Reply by mprithiv** [reply]
Hi. It was my mistake. There was a type (indes instead of index). Now I am able to convert but however `RandomNodeSplit` based on `key=None` doesn't still seem to work
`Data(node_type=[8697], edge_index=[2, 1343836], edge_type=[1343836])`
With `Random` split,
`nodesplit = RandomNodeSplit(split='random', key=None)
nodesplit(transformed_homodata)
`
there is a type error. 
From [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.RandomNodeSplit), does this mean the node split will work only with a key (possibly y parameter). If that is the case, how does the GAT training on hetero graph work for unsupervised approach

**Reply by rusty1s** [reply]
The following works for me:
```python
from torch_geometric.datasets import FakeHeteroDataset
from torch_geometric.transforms import RandomNodeSplit

dataset = FakeHeteroDataset()
data = dataset[0]
print(data)

transform = RandomNodeSplit(num_val=0.2, num_test=0.2, key=None)
data = transform(data)
print(data)
```
Can you confirm?

**Reply by mprithiv** [reply]
Trying out right away.

**Reply by mprithiv** [reply]
I tried on both Heterogeneous and Homogenous data. 

Before NodeSplit on Homogenous data
==============================

`Data(node_type=[8697], edge_index=[2, 671918], edge_type=[671918])`

After NodeSplit on Homogenous data
==============================
`Data(node_type=[8697], edge_index=[2, 671918], edge_type=[671918], train_mask=[8697], val_mask=[8697], test_mask=[8697])`

On Heterogeneous

Original Data
====================
`HeteroData(
  buyer={
    num_nodes=7813,
    x=[7813, 1152]
  },
  seller={
    num_nodes=884,
    x=[884, 387]
  },
  (buyer, txn_amount, seller)={ edge_index=[2, 335959] },
  (buyer, txn_date, seller)={ edge_index=[2, 335959] }
)`

Transformed Data
====================
`HeteroData(
  buyer={
    num_nodes=7813,
    x=[7813, 1152],
    train_mask=[7813],
    val_mask=[7813],
    test_mask=[7813]
  },
  seller={
    num_nodes=884,
    x=[884, 387],
    train_mask=[884],
    val_mask=[884],
    test_mask=[884]
  },
  (buyer, txn_amount, seller)={ edge_index=[2, 335959] },
  (buyer, txn_date, seller)={ edge_index=[2, 335959] }
)
`


Does this supposed to split 7813/884 (or 8697 if homogenous) into 60-20-20? Or is it because of `key=None`, utilizes all the data for train, test and validation?

**Reply by rusty1s** [reply]
Can you clarify what you mean? You can specify `num_val` and `num_test` to denote how many nodes you want to use for validation and testing, respectively. The remaining nodes are then used for training.

**Reply by mprithiv** [reply]
I have specified num_val and num_test as 0.2. With that, I see all the train_mask, test_mask, val_mask as same [7813] which is the num nodes for buyer node type.

**Reply by rusty1s** [reply]
Yes, this is a mask: a boolean vector with a shape equal to the number of nodes, in which only a certain fraction of values are set to `True`.
